### ISSUE TITLE
[App]: hardware logging in multi-node setting

### ISSUE THREAD

**Issue by BramVanroy** [wandb/wandb#7470]
### Current Behavior

Currently, in th e run overview, we can get an idea of the system hardware, specifically GPU count and CPU count. However, as far as I can tell this does not account for multi-node settings and only reports what the current node is equipped with. While I understand why that is the case, it may be confusing because it is not "correct".

### Expected Behavior

Correct hardware information. To be honest I am not sure how feasible it is to collect this information without integration with distributed communication frameworks or something else custom.

### Environment

OS: Linux

Browsers: Edge

**Comment by thanos-wandb** [comment]
Hi @BramVanroy thank you for reporting this. May I please ask some more context, what's your current compute infra? and which ML frameworks are you mostly using?

**Comment by thanos-wandb** [comment]
Hi @BramVanroy just following up on this, to see if you could provide us with some additional information on your current multinode infrastructure so as to include those in a feature request for our engineers? thank you!

**Comment by BramVanroy** [comment]
Hi Thanos

I am running jobs on between 1 node, 1 GPU up to 10 nodes, 4 GPUs each. It seems to me that wandb does not correctly log hardware when it comes to multi-node settings.

**Comment by thanos-wandb** [comment]
Perfect, thank you @BramVanroy for the additional context. I was wondering what's reported in those runs, if you navigate in `Files` view and open the `wandb-metadata.json` file, under the  "cpu_count" and "cpu_count_logical" entries. Does it not detect the correct hw info when it's multinode?

**Comment by BramVanroy** [comment]
Correct. It only reports the main node hardware configuration, but not the whole pool.

**Comment by thanos-wandb** [comment]
Great, thank you @BramVanroy for the clarification. I have logged this feature request with our engineers, and we will keep you updated here on any progress.

**Comment by nepfaff** [comment]
This would be useful to have

**Comment by abraham-leal** [comment]
For Nvidia GPUs, this may be a good workaround on this issue: https://wandb.ai/dimaduev/dcgm/reports/Monitoring-GPU-cluster-performance-with-NVIDIA-DCGM-Exporter-and-Weights-Biases--Vmlldzo0MDYxMTA1

**Comment by nima-rezaeian-ai** [comment]
This is something we're interested in as well, also seems to be duplicate of https://github.com/wandb/wandb/issues/7324. If there was a way for the wandb sdk to attach to an existing run and log only system metrics I think that'd cover our use case.

**Comment by thanos-wandb** [comment]
Hi @nima-rezaeian-ai thank you for reaching out - would it be possible to provide more information on your use case? are you using our HuggingFace integration for your distributed ML training, as in the Github issue you've linked? it should be possible to log system metrics by starting a run in each process, and then use the `group` keyword to have them grouped as described here: https://docs.wandb.ai/guides/track/log/distributed-training/#method-2-many-processes

**Comment by nepfaff** [comment]
Hey @noaleetz, I'm unfortunately not available for a chat at the moment but I can give you a quick answer to these questions.

### Frameworks you're using for distributed training
I'm using wandb through pytorch lightning. I'm using torchrun for distributed training.

### How are distributed training jobs spun up in your setup
At the moment, I'm using multiple AWS machines through EC2 and then manually run torchrun on each of these. In the future, I'm planning to switch to AWS Sagemaker but I didn't get to it yet.

### Network topology and what is visible from main node
Different AWS nodes. All of them have an internet connection.

I'd love to use wandb for the following:
- discover hardware issues on specific nodes
- track the used resources with all my other training logs
- help me optimize resource utilization across multiple nodes

**Comment by noaleetz** [comment]
> #7324

Hi @nima-rezaeian-ai - thanks for writing in with some of this info. When you say "wandb sdk to attach to an existing run" - you mean for a node where the run was not initialized in, you would like for system metrics to also be logged to the existing run (i.e the run initialized via wandb.init() in the master node)? Rather than initializing 1 run per node, for example?

**Comment by farzadab** [comment]
Hi Noa,
Not sure where you are with this since it's been a while, but here's my take on this:

> And when you say "log only system metrics" - you're pointing out here that performance metrics don't need to be tracked across nodes

Yes. For a multi-node run, things like `train/loss` or `eval/perplexity` will usually not need to be logged separately. What I did expect to be different though would be metrics related to the system. Here are a few based on my [recent W&B run](https://wandb.ai/fixie/ultravox/runs/7mkqpb6h) (not a public link, but I assumed you could get access):

In run information I expected to see either 16 GPUs or 2x8 GPUs being acknowledged:
<img width="456" alt="Image" src="https://github.com/user-attachments/assets/af9af60e-d0bd-4996-bd94-a71c2f295798" />

And to see the utilization metrics for all GPUs (either all in one plot or separated into two plots: GPU Utilization Node 0 vs 1):
<img width="695" alt="Image" src="https://github.com/user-attachments/assets/339a9a2e-42f0-4dd6-9c60-0c557abf48cb" />

A bonus would be to get the logs for each machine separately, probably only for process_0 of each node.

**Comment by noaleetz** [comment]
Hi folks! Thanks for all the feedback here - we've been working on a bunch of stuff in this area in the past months and wanted to share an update. 

These are still in an experimental state but would be awesome if you want to try it out and share feedback. 

Speaking specifically about these requests:
- system metric plots for each individual processes
- console logs for each individual processes

**New Run Settings:**
- Shared Mode: Multiple processes can now log to a single run without conflicts - console logs and system metrics are collected from each node you call wandb.init on
- Process Labels (x_label): In each process you call wandb.init, you can define a label (i.e rank0) which later will allow you to filter the console logs and system metrics by process 

**Outcome** - you can log metrics from different processes/writers to a single run, and you'll see that now console logs and system metrics will be grouped by the relevant label

`# spinning up a run inside an individual node with rank-{global_rank}
settings = wandb.Settings(
        mode="shared",
        x_label=f"rank-{global_rank}",
)

run = wandb.init(
        id=SHARED_RUN_ID,
        settings=settings,
    )
`

Note: you will need to be on SDK version 0.19.5 or greater.  

To see the full workflow - checkout [this](https://wandb.ai/dimaduev/simple-cnn-ddp/reports/Distributed-Training-with-Shared-Mode--VmlldzoxMTI0NTE1NA) demo report.

**Comment by noaleetz** [comment]
@farzadab At the moment the System Hardware shown in run overview page still just reflects the overview for a single node, but it's on our roadmap to merge this. 

For now though, you should be able to:
- see logs for each machine separately
- system metrics for each processes appear as their own plot

<img width="779" alt="Image" src="https://github.com/user-attachments/assets/44e00769-56c7-4f47-9ee6-4662432e5829" />

<img width="1195" alt="Image" src="https://github.com/user-attachments/assets/f526937a-49d1-496a-97b2-90b478907f3a" />

**Comment by ntenenz** [comment]
@noaleetz, is there an upper bound on the number of system panels that can be displayed/logged? I implemented something that roughly looks like this

```python
if LOCAL_RANK == 0: 
    is_rk0 = RANK == 0   
    wandb.init(
        config=config_dict,
        settings=wandb.Settings(
            mode="shared",
            x_label=f"rank-{RANK}",
            x_primary=is_rk0,
            x_update_finish_state=is_rk0,
        ),
    )
```

but discovered many of the system panels were missing. Suspiciously, the total number of available system panels is 256.

If 256 is a fixed upper bound, is there a way to disable less relevant panels so that the higher priority ones will be displayed?

**Comment by thanos-wandb** [comment]
Hi @ntenenz thank you for following-up on this. In order to control which ranks to be logged, you could make use of the following setting option, eg if higher priority for you are the first 4: `x_stats_gpu_device_ids=(0, 1, 2, 3)`. Let me know if that would work for you, and regarding the amount of system panels, would it be please possible to share a screenshot to understand this issue better? are you referring to how many system panels can exist in the systems section, or how many lines could appear in one system panel? thanks!

**Comment by ntenenz** [comment]
The primary ask is whether more than 256 system metrics, totaled across all ranks, can be logged. If not, can this be made possible?

Right now, it appears the metrics are sorted by name then rank and then the list is truncated. Concretely, if given a world size of 128 and 4 system metrics per rank (metric1 through metric4), only metric1 and metric2 will be visible.

If logging more than 256 system metrics is not possible, as a temporary placeholder, at least being able to control which metric across all ranks is logged would be beneficial. In the above scenario, this could correspond to logging metric3 and metric4 for all ranks.

**Comment by thanos-wandb** [comment]
Hi @ntenenz could you please share a link to your workspace and the SDK debug logs to investigate this further? those can be found in local wandb/run-timestamp-id/logs directory. All system metrics should be logged, however for performance reasons the amount of those displayed in the app are capped.

**Comment by ntenenz** [comment]
Is this cap configurable? For even small-scale, multinode jobs, the number of system metrics (when logged in shared mode) will quicly exceed 256.

**Comment by thanos-wandb** [comment]
Hi @ntenenz more than happy to look into it, however, I am not sure why the total number of system metrics will change. The same system metrics will be captured, eg Disk Utilization, GPU Power Usage, etc. Therefore, that still takes me back to my previous questions, and if the ask here is about how many panels you are seeing under the `System` section, or how many line plots are rendered in each one of them. Based on your previous example, you should only be able to see 4 panels (metric1 through metric4) and for each panel to have the 128 world_size ranks line plots rendered. A screenshot as requested earlier would be really useful. Also, if you could share if that run has a successful `Finished` state, or if it `Crashed`/`Failed`, that would be much appreciated too to help further troubleshooting. Additionally, I am wondering what happens when you try to Add a new panel in this section, and search for the key in y-axis, does it show up there?

**Comment by thanos-wandb** [comment]
Hi @ntenenz I wanted to check back in on this issue as we would like to resolve this for you, would it be please possible to reach out via support@wandb.com? email is the best place to exchange any sensitive details we’ll need for troubleshooting. If you could share previously requested information, and the debug logs from SDK for this affected run, it would be really helpful. Thank you!
